Sources/Controllers/Map/Layers/OAGPXLayer.mm

```diff
@@ -244,12 +244,24 @@ - (void)configureCachedWallColorsFor:(EOAGradientScaleType)type
     }
     else
     {
-        OARouteColorizationHelper *routeColorization =
-        [[OARouteColorizationHelper alloc] initWithGpxFile:doc
-                                                  analysis:analysis?:[doc getAnalysis:0]
-                                                      type:[OAGradientScaleType withGradientScaleType:type].toColorizationType
-                                           maxProfileSpeed:0];
-        _cachedWallColors[key] = routeColorization ? [routeColorization getResult] : QList<OsmAnd::FColorARGB>();
+        ColorPalette *palette =
+            [[ColorPaletteHelper shared] getGradientColorPaletteSync:[type toColorizationType]
+                                                 gradientPaletteName:_gradientPalette
+                                                             refresh:NO];
+        OARouteColorize *routeColorize =
+            [[OARouteColorize alloc] initWithGpxFile:doc
+                                            analysis:analysis ?: [doc getAnalysis:0]
+                                                type:[type toColorizationType]
+                                             palette:palette
+                                     maxProfileSpeed:0];
+        _cachedWallColors[key].clear();
+        if (routeColorize)
+        {
+            for (OARouteColorizationPoint *colorizationPoint in [routeColorize getResult])
+            {
+                _cachedWallColors[key].append(OsmAnd::ColorARGB(colorizationPoint.color));
```

Comment:
Why not just return QList of colors immediately?
 _cachedWallColors[key] =  [routeColorize colors];